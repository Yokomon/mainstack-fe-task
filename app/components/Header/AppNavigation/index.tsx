"use client";

import React, { Fragment, useMemo, useState } from "react";
import { IoIosArrowForward } from "@react-icons/all-files/io/IoIosArrowForward";
import { Menu, Transition } from "@headlessui/react";
import clsx from "clsx";
import Image from "next/image";

import { appsDropdown } from "@/app/constants/navbar";
import { HEADER_TITLE, getHeaderNavs } from "@/app/utils";
import { MdKeyboardArrowDown } from "@react-icons/all-files/md/MdKeyboardArrowDown";
import Link from "next/link";

const AppNavigation = () => {
  const [navPath, setNavPath] = useState<keyof typeof HEADER_TITLE>("Revenue");

  const headerNavs = useMemo(() => getHeaderNavs(navPath), [navPath]);
  return (
    <div className="flex items-center space-x-2 xl:ml-32">
      {headerNavs.map(({ headerTitle, icon, headerImage, active }) => (
        <Link
          href={"/"}
          onClick={() => setNavPath(headerTitle)}
          key={headerTitle}
          className={clsx({
            ["p-2.5 flex items-center space-x-2 duration-500 cursor-pointer rounded-full px-5 text-sm "]:
              true,
            ["bg-black p-3 rounded-full text-white"]: active,
            ["hover:bg-gray-100 hover:text-gray-600"]: !active,
          })}
        >
          {icon ? React.createElement(icon, { size: 18 }) : headerImage}
          <h3>{headerTitle}</h3>
        </Link>
      ))}
      <div className="relative">
        <Menu>
          {({ open }) => (
            <>
              <Menu.Button
                className={clsx({
                  ["p-2.5 flex items-center space-x-2 duration-500  rounded-full px-5 text-sm"]:
                    true,
                  ["bg-black text-white"]: open,
                  ["hover:bg-gray-100"]: !open,
                })}
              >
                <Image
                  src="/images/navbar_assets/widgets.svg"
                  width={20}
                  height={20}
                  alt="widgets"
                  className={clsx({
                    ["brightness-[100]"]: open,
                  })}
                />
                <h3>Apps</h3>
                {open ? (
                  <div
                    className={clsx({
                      ["opacity-100 !flex items-center space-x-3"]: true,
                    })}
                  >
                    <h3 className="text-sm pl-3">Link in Bio</h3>
                    <MdKeyboardArrowDown size={18} />
                  </div>
                ) : null}
              </Menu.Button>
              <Transition
                as={Fragment}
                enter="transition ease-out duration-300"
                enterFrom="transform opacity-0 scale-95"
                enterTo="transform opacity-100 scale-100"
                leave="transition ease-in duration-100"
                leaveFrom="transform opacity-100 scale-100"
                leaveTo="transform opacity-0 scale-95"
              >
                <Menu.Items className="absolute space-y-3 left-0 mt-3 w-96 pb-8 origin-top-left rounded-lg bg-white shadow-xl p-3 ring-1 ring-black/5 focus:outline-none">
                  {appsDropdown.map(
                    ({ appTitle, appSubtitle, appIcon, appImage }) => (
                      <Menu.Item key={appTitle}>
                        <button className="flex group items-center space-x-5 hover:ring-1 ring-gray-100 pr-10 p-2 py-4 rounded-xl w-full hover:shadow-md duration-500">
                          <div className="rounded-lg border border-gray-100 p-2">
                            {appImage ? (
                              appImage
                            ) : (
                              <div className="p-1.5">
                                {React.createElement(appIcon, {
                                  size: 22,
                                  className: "text-sky-600",
                                })}
                              </div>
                            )}
                          </div>
                          <div className="flex flex-1 flex-col items-start">
                            <h2>{appTitle}</h2>
                            <p className="text-sm text-gray-500">
                              {appSubtitle}
                            </p>
                          </div>
                          <IoIosArrowForward className="opacity-0 duration-500 group-hover:opacity-100" />
                        </button>
                      </Menu.Item>
                    )
                  )}
                </Menu.Items>
              </Transition>
            </>
          )}
        </Menu>
      </div>
    </div>
  );
};

export default AppNavigation;
