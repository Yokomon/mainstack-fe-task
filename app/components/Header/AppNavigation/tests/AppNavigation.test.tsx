import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import AppNavigation from "..";

describe("AppNavigation Component", () => {
  it("renders without crashing", () => {
    render(<AppNavigation />);
    expect(screen.getByText("Revenue")).toBeInTheDocument();
  });

  it("renders each navigation links with the correct text", () => {
    render(<AppNavigation />);
    const links = screen.getAllByRole("link");
    const linkText = links.map((button) => button.textContent);
    expect(linkText).toEqual(["Home", "Analytics", "Revenue", "CRM"]);
  });

  it('opens the Apps dropdown when the "Apps" button is clicked', () => {
    render(<AppNavigation />);
    const appsButton = screen.getByText("Apps");
    fireEvent.click(appsButton);
    expect(screen.getByRole("menu")).toBeInTheDocument();
  });
  it("fires the correct action when an app in the dropdown is clicked", () => {
    render(<AppNavigation />);
    const appsLink = screen.getByText("Apps");
    fireEvent.click(appsLink);
    const modalText = screen.getByTestId("bio-link");
    expect(modalText).toBeInTheDocument();
  });

  it("closes the dropdown when clicking outside of it", () => {
    render(<AppNavigation />);
    const appsButton = screen.getByText("Apps");
    fireEvent.click(appsButton);
    const outsideElement = screen.getByTestId("bio-link");
    fireEvent.click(appsButton);
    expect(outsideElement).not.toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<AppNavigation />);
    expect(asFragment()).toMatchSnapshot();
  });
});
