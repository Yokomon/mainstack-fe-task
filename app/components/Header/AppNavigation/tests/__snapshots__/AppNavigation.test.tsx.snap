// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppNavigation Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center space-x-2 xl:ml-32"
  >
    <a
      class="p-2.5 flex items-center space-x-2 duration-500 cursor-pointer rounded-full px-5 text-sm  hover:bg-gray-100 hover:text-gray-600"
      href="/"
    >
      <svg
        fill="currentColor"
        height="18"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1016.7 513.36L536.331 10.192a31.924 31.924 0 0 0-23.088-9.84 32.038 32.038 0 0 0-23.088 9.84L7.307 513.344c-12.24 12.752-11.808 32.992.944 45.248 12.752 12.224 32.992 11.872 45.248-.944l43.007-44.832v478.832c0 17.68 14.336 32 32 32h223.552c17.632 0 31.936-14.256 32-31.905l1.008-319.664h254.992v319.568c0 17.68 14.32 32 32 32H895.53c17.68 0 32-14.32 32-32V512.655l42.992 45.04a31.997 31.997 0 0 0 23.09 9.84c7.967 0 15.967-2.944 22.16-8.944 12.736-12.224 13.152-32.48.928-45.232zm-153.165-58.544v504.831H704.063V640.095c0-17.68-14.32-32-32-32h-318.88c-17.632 0-31.936 14.256-32 31.904l-1.008 319.664H160.511V454.815c0-2.64-.416-5.168-1.008-7.632L513.263 78.56l351.424 368.208c-.688 2.592-1.152 5.264-1.152 8.048z"
        />
      </svg>
      <h3>
        Home
      </h3>
    </a>
    <a
      class="p-2.5 flex items-center space-x-2 duration-500 cursor-pointer rounded-full px-5 text-sm  hover:bg-gray-100 hover:text-gray-600"
      href="/"
    >
      <svg
        fill="currentColor"
        height="18"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14z"
        />
        <path
          d="M7 12h2v5H7zM15 7h2v10h-2zM11 14h2v3h-2zM11 10h2v2h-2z"
        />
      </svg>
      <h3>
        Analytics
      </h3>
    </a>
    <a
      class="p-2.5 flex items-center space-x-2 duration-500 cursor-pointer rounded-full px-5 text-sm  bg-black p-3 rounded-full text-white"
      href="/"
    >
      <img
        alt="payments-image"
        class=""
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/images/navbar_assets/payments.svg"
        style="color: transparent;"
        width="20"
      />
      <h3>
        Revenue
      </h3>
    </a>
    <a
      class="p-2.5 flex items-center space-x-2 duration-500 cursor-pointer rounded-full px-5 text-sm  hover:bg-gray-100 hover:text-gray-600"
      href="/"
    >
      <img
        alt="group-image"
        class=""
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/images/navbar_assets/group.svg"
        style="color: transparent;"
        width="20"
      />
      <h3>
        CRM
      </h3>
    </a>
    <div
      class="relative"
    >
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="p-2.5 flex items-center space-x-2 duration-500  rounded-full px-5 text-sm hover:bg-gray-100"
        data-headlessui-state=""
        id="headlessui-menu-button-:rn:"
        type="button"
      >
        <img
          alt="widgets"
          class=""
          data-nimg="1"
          decoding="async"
          height="20"
          loading="lazy"
          src="/images/navbar_assets/widgets.svg"
          style="color: transparent;"
          width="20"
        />
        <h3>
          Apps
        </h3>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
